### PR TITLE
ENG-472: Add order and transfer without fees encoding

### PR DESCRIPTION
**Pull request by JoseAfonsoThreeSigma**
## Describe your changes
Add order and transfer without fees encoding

## Checklist before requesting a review
- [X] I have performed a self-review of my code
- [X] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.

## Reviewers
@JoseAfonsoThreeSigma @asynctomatic

### CODE REVIEW

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs [hunk 1]
```diff
@@ -55,6 +55,29 @@ public BigInteger EncodeLimitOrderWithFees(EncodeLimitOrderWithFeesModel encodeL
         return pedersenHash.CreateHash(thirdInnerHash, fifthWeight);
     }
 
+    /// <inheritdoc />
+    public BigInteger EncodeLimitOrderWithoutFees(EncodeLimitOrderWithoutFeesModel encodeLimitOrderWithoutFeesModel)
```
Comment (asynctomatic) [inline src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs:59]: This is functionally equivalent to `DeprecatedHashLimitOrder`.

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs [hunk 2]
```diff
@@ -88,21 +111,79 @@ public BigInteger EncodeTransferWithFees(EncodeTransferWithFeesModel encodeTrans
         return pedersenHash.CreateHash(thirdInnerHash, fifthWeight);
     }
 
+    /// <inheritdoc />
+    public BigInteger EncodeTransferWithoutFees(EncodeTransferWithoutFeesModel encodeTransferWithoutFeesModel)
```
Comment (asynctomatic) [inline src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs:115]: This is functionally equivalent to `DeprecatedHashTransferOrder`.

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs [hunk 3]
```diff
@@ -88,21 +111,79 @@ public BigInteger EncodeTransferWithFees(EncodeTransferWithFeesModel encodeTrans
...
-        var condition = pedersenHash.CreateHash(
-            new BigInteger(encodeTransferWithFeesModel.Fact.RemoveHexPrefix(), 16),
-            new BigInteger(encodeTransferWithFeesModel.FactRegistryAddress.RemoveHexPrefix(), 16));
+        var condition = GetCondition(encodeTransferWithFeesModel.Fact, encodeTransferWithFeesModel.FactRegistryAddress);
 
         return EncodeConditionalTransferWithFees(encodeTransferWithFeesModel, condition.ToByteArray().ToHex());
     }
 
+    /// <inheritdoc />
+    public BigInteger EncodeConditionalTransferWithoutFees(EncodeTransferWithoutFeesModel encodeTransferWithoutFeesModel)
```
Comment (asynctomatic) [inline src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs:157]: This is functionally equivalent to `DeprecatedHashConditionalTransfer`.

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs [hunk 4]
```diff
@@ -141,6 +222,35 @@ public BigInteger EncodeConditionalTransferWithFees(EncodeTransferWithFeesModel
         return pedersenHash.CreateHash(fourthInnerHash, fifthWeight);
     }
 
+    public BigInteger EncodeConditionalTransferWithoutFees(EncodeTransferWithoutFeesModel encodeTransferWithoutFeesModel, string condition)
```
Comment (asynctomatic) [inline src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs:225]: What was the reason for overloading the `EncodeConditionalTransferWithoutFees` method to explicitly accept a condition? The `EncodeConditionalTransferWithoutFees` is also being overloaded in the same way.

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs [hunk 5]
```diff
@@ -88,21 +111,79 @@ public BigInteger EncodeTransferWithFees(EncodeTransferWithFeesModel encodeTrans
...
+            new BigInteger(encodeTransferWithoutFeesModel.AssetIdSold.RemoveHexPrefix(), 16),
+            new BigInteger(encodeTransferWithoutFeesModel.ReceiverStarkKey.RemoveHexPrefix(), 16));
+
+        return pedersenHash.CreateHash(firstInnerHash, packedMessage);
+    }
+
     /// <inheritdoc />
     public BigInteger EncodeConditionalTransferWithFees(EncodeTransferWithFeesModel encodeTransferWithFeesModel)
     {
+        Guards.NotNullOrEmptyOrWhitespace(encodeTransferWithFeesModel.AssetIdSold);
```
Comment (asynctomatic) [inline src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs:140]: Shouldn't we validate this at creation of the model object (ie. `EncodeTransferWithFeesModel`), following a similar approach to a regular value object? I don't see a reason why we would ever initialise this model in an invalid state and moving the validation would make this code less verbose.

If decide on moving forward with this we should do it on a separate PR.

**Review by asynctomatic — CHANGES_REQUESTED** [review]
Most of the changes introduced in this PR seem to be duplicates of already implemented methods (`Deprecated` prefix), with the addition of using model objects instead of explicitly passing the parameters to encode.

I agree with changes to use model objects for consistency with the remaining methods, but I think we would further benefit from their usage if we moved the validation (ie. `Guard` logic) to the model creation itself (left a comment on this).

Additionally, I believe we should keep the `Obsolete` annotations for these methods without fees since these transaction types seem to have been deprecated in the latest StarkEx versions. I believe setting the fee amount to zero is how txs without fees should be encoded but we'll confirm this once we get an answer from the Starkware team.

**Comment by JoseAfonsoThreeSigma** [comment]
> Most of the changes introduced in this PR seem to be duplicates of already implemented methods (`Deprecated` prefix), with the addition of using model objects instead of explicitly passing the parameters to encode.
> 
> I agree with changes to use model objects for consistency with the remaining methods, but I think we would further benefit from their usage if we moved the validation (ie. `Guard` logic) to the model creation itself (left a comment on this).
> 
> Additionally, I believe we should keep the `Obsolete` annotations for these methods without fees since these transaction types seem to have been deprecated in the latest StarkEx versions. I believe setting the fee amount to zero is how txs without fees should be encoded but we'll confirm this once we get an answer from the Starkware team.

My bad, didn't realize this implementation is similar to the existing respective deprecated implementations in this SDK. Disregard the changes on this PR for now, still waiting on an answer from StarkWare on how message encodings should be calculated when no fees are applied

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs (hunk 1 above)
Comment (JoseAfonsoThreeSigma) [reply src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs]: Deleted the existing deprecated method in favor of this one

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs (hunk 2 above)
Comment (JoseAfonsoThreeSigma) [reply src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs]: Deleted the existing deprecated method in favor of this one

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs (hunk 3 above)
Comment (JoseAfonsoThreeSigma) [reply src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs:205]: Deleted the existing deprecated method in favor of this one

File: src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs (hunk 4 above)
Comment (JoseAfonsoThreeSigma) [reply src/StarkEx.Crypto.SDK/Hashing/SpotTradingMessageHasher.cs]: As talked in daily, it was to be able to use the same test dataset from StarkWare which only includes the condition

**Review by asynctomatic — APPROVED** [review]
